File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/scheduling/NotificationScheduler.java
package expo.modules.notifications.notifications.scheduling;

import android.content.Context;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.os.ResultReceiver;

import expo.modules.core.ExportedModule;
import expo.modules.core.Promise;
import expo.modules.core.arguments.ReadableArguments;
import expo.modules.core.errors.InvalidArgumentException;
import expo.modules.core.interfaces.ExpoMethod;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Date;

import androidx.annotation.Nullable;
import expo.modules.notifications.notifications.ArgumentsNotificationContentBuilder;
import expo.modules.notifications.notifications.NotificationSerializer;
import expo.modules.notifications.notifications.interfaces.NotificationTrigger;
import expo.modules.notifications.notifications.interfaces.SchedulableNotificationTrigger;
import expo.modules.notifications.notifications.model.NotificationContent;
import expo.modules.notifications.notifications.model.NotificationRequest;
import expo.modules.notifications.notifications.triggers.ChannelAwareTrigger;
import expo.modules.notifications.notifications.triggers.DailyTrigger;
import expo.modules.notifications.notifications.triggers.DateTrigger;
import expo.modules.notifications.notifications.triggers.TimeIntervalTrigger;
import expo.modules.notifications.notifications.triggers.WeeklyTrigger;
import expo.modules.notifications.notifications.triggers.YearlyTrigger;
import expo.modules.notifications.service.NotificationsService;

public class NotificationScheduler extends ExportedModule {
  private final static String EXPORTED_NAME = "ExpoNotificationScheduler";

  protected static Handler HANDLER = new Handler(Looper.getMainLooper());

  public NotificationScheduler(Context context) {
    super(context);
  }

  @Override
  public String getName() {
    return EXPORTED_NAME;
  }

  protected Context getSchedulingContext() {
    return getContext();
  }

  @ExpoMethod
  public void getAllScheduledNotificationsAsync(final Promise promise) {
    NotificationsService.Companion.getAllScheduledNotifications(getSchedulingContext(), new ResultReceiver(HANDLER) {
      @Override
      protected void onReceiveResult(int resultCode, Bundle resultData) {
        super.onReceiveResult(resultCode, resultData);
        if (resultCode == NotificationsService.SUCCESS_CODE) {
          Collection<NotificationRequest> requests = resultData.getParcelableArrayList(NotificationsService.NOTIFICATION_REQUESTS_KEY);
          if (requests == null) {
            promise.reject("ERR_NOTIFICATIONS_FAILED_TO_FETCH", "Failed to fetch scheduled notifications.");
          } else {
            promise.resolve(serializeScheduledNotificationRequests(requests));
          }
        } else {
          Exception e = resultData.getParcelable(NotificationsService.EXCEPTION_KEY);
          promise.reject("ERR_NOTIFICATIONS_FAILED_TO_FETCH", "Failed to fetch scheduled notifications.", e);
        }
      }
    });
  }

  @ExpoMethod
  public void scheduleNotificationAsync(final String identifier, ReadableArguments notificationContentMap, ReadableArguments triggerParams, final Promise promise) {
    try {
      NotificationContent content = new ArgumentsNotificationContentBuilder(getContext()).setPayload(notificationContentMap).build();
      NotificationRequest request = createNotificationRequest(identifier, content, triggerFromParams(triggerParams));
      NotificationsService.Companion.schedule(getSchedulingContext(), request, new ResultReceiver(HANDLER) {
          @Override
          protected void onReceiveResult(int resultCode, Bundle resultData) {
            super.onReceiveResult(resultCode, resultData);
            if (resultCode == NotificationsService.SUCCESS_CODE) {
              promise.resolve(identifier);
            } else {
              Exception e = resultData.getParcelable(NotificationsService.EXCEPTION_KEY);
              if (e == null) {
                promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule notification.");
              } else {
                promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. " + e.getMessage(), e);
              }
            }
          }
        });
    } catch (InvalidArgumentException e) {
      promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. " + e.getMessage(), e);
    } catch (NullPointerException e) {
      promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. Encountered unexpected null value. " + e.getMessage(), e);
    }
  }

  @ExpoMethod
  public void cancelScheduledNotificationAsync(String identifier, final Promise promise) {
    NotificationsService.Companion.removeScheduledNotification(getSchedulingContext(), identifier, new ResultReceiver(HANDLER) {
      @Override
      protected void onReceiveResult(int resultCode, Bundle resultData) {
        super.onReceiveResult(resultCode, resultData);
        if (resultCode == NotificationsService.SUCCESS_CODE) {
          promise.resolve(null);
        } else {
          Exception e = resultData.getParcelable(NotificationsService.EXCEPTION_KEY);
          promise.reject("ERR_NOTIFICATIONS_FAILED_TO_CANCEL", "Failed to cancel notification.", e);
        }
      }
    });
  }

  @ExpoMethod
  public void cancelAllScheduledNotificationsAsync(final Promise promise) {
    NotificationsService.Companion.removeAllScheduledNotifications(getSchedulingContext(), new ResultReceiver(HANDLER) {
      @Override
      protected void onReceiveResult(int resultCode, Bundle resultData) {
        super.onReceiveResult(resultCode, resultData);
        if (resultCode == NotificationsService.SUCCESS_CODE) {
          promise.resolve(null);
        } else {
          Exception e = resultData.getParcelable(NotificationsService.EXCEPTION_KEY);
          promise.reject("ERR_NOTIFICATIONS_FAILED_TO_CANCEL", "Failed to cancel all notifications.", e);
        }
      }
    });
  }

  @ExpoMethod
  public void getNextTriggerDateAsync(ReadableArguments triggerParams, Promise promise) {
    try {
      NotificationTrigger trigger = triggerFromParams(triggerParams);
      if (trigger instanceof SchedulableNotificationTrigger) {
        Date nextTriggerDate = ((SchedulableNotificationTrigger) trigger).nextTriggerDate();
        if (nextTriggerDate == null) {
          promise.resolve(null);
        } else {
          promise.resolve((double) (nextTriggerDate.getTime()));
        }
      } else {
        String triggerDescription = trigger == null ? "null" : trigger.getClass().getName();
        String message = String.format("It is not possible to get next trigger date for triggers other than calendar-based. Provided trigger resulted in %s trigger.", triggerDescription);
        promise.reject("ERR_NOTIFICATIONS_INVALID_CALENDAR_TRIGGER", message);
      }
    } catch (InvalidArgumentException e) {
      promise.reject("ERR_NOTIFICATIONS_FAILED_TO_GET_NEXT_TRIGGER_DATE", "Failed to get next trigger date for the trigger. " + e.getMessage(), e);
    } catch (NullPointerException e) {
      promise.reject("ERR_NOTIFICATIONS_FAILED_TO_GET_NEXT_TRIGGER_DATE", "Failed to get next trigger date for the trigger. Encountered unexpected null value. " + e.getMessage(), e);
    }
  }

  @Nullable
  protected NotificationTrigger triggerFromParams(@Nullable ReadableArguments params) throws InvalidArgumentException {
    if (params == null) {
      return null;
    }

    String channelId = params.getString("channelId", null);
    switch (params.getString("type")) {
      case "timeInterval":
        if (!(params.get("seconds") instanceof Number)) {
          throw new InvalidArgumentException("Invalid value provided as interval of trigger.");
        }
        return new TimeIntervalTrigger(((Number) params.get("seconds")).longValue(), params.getBoolean("repeats"), channelId);
      case "date":
        if (!(params.get("timestamp") instanceof Number)) {
          throw new InvalidArgumentException("Invalid value provided as date of trigger.");
        }
        return new DateTrigger(((Number) params.get("timestamp")).longValue(), channelId);
      case "daily":
        if (!(params.get("hour") instanceof Number) || !(params.get("minute") instanceof Number)) {
          throw new InvalidArgumentException("Invalid value(s) provided for daily trigger.");
        }
        return new DailyTrigger(
          ((Number) params.get("hour")).intValue(),
          ((Number) params.get("minute")).intValue(),
          channelId
        );
      case "weekly":
        if (!(params.get("weekday") instanceof Number) || !(params.get("hour") instanceof Number) || !(params.get("minute") instanceof Number)) {
          throw new InvalidArgumentException("Invalid value(s) provided for weekly trigger.");
        }
        return new WeeklyTrigger(
          ((Number) params.get("weekday")).intValue(),
          ((Number) params.get("hour")).intValue(),
          ((Number) params.get("minute")).intValue(),
          channelId
        );
      case "yearly":
        if (!(params.get("day") instanceof Number) || !(params.get("month") instanceof Number) || !(params.get("hour") instanceof Number) || !(params.get("minute") instanceof Number)) {
          throw new InvalidArgumentException("Invalid value(s) provided for yearly trigger.");
        }
        return new YearlyTrigger(
          ((Number) params.get("day")).intValue(),
          ((Number) params.get("month")).intValue(),
          ((Number) params.get("hour")).intValue(),
          ((Number) params.get("minute")).intValue(),
          channelId
        );
      case "channel":
        return new ChannelAwareTrigger(channelId);
      default:
        throw new InvalidArgumentException("Trigger of type: " + params.getString("type") + " is not supported on Android.");
    }
  }

  protected NotificationRequest createNotificationRequest(String identifier, NotificationContent content, NotificationTrigger notificationTrigger) {
    return new NotificationRequest(identifier, content, notificationTrigger);
  }

  protected Collection<Bundle> serializeScheduledNotificationRequests(Collection<NotificationRequest> requests) {
    Collection<Bundle> serializedRequests = new ArrayList<>(requests.size());
    for (NotificationRequest request : requests) {
      serializedRequests.add(NotificationSerializer.toBundle(request));
    }
    return serializedRequests;
  }
}

File: packages/expo-notifications/android/src/main/java/expo/modules/notifications/notifications/scheduling/NotificationScheduler.kt
package expo.modules.notifications.notifications.scheduling

import android.content.Context
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import expo.modules.core.arguments.ReadableArguments
import expo.modules.core.errors.InvalidArgumentException
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.notifications.ResultReceiverBody
import expo.modules.notifications.createDefaultResultReceiver
import expo.modules.notifications.notifications.ArgumentsNotificationContentBuilder
import expo.modules.notifications.notifications.NotificationSerializer
import expo.modules.notifications.notifications.interfaces.NotificationTrigger
import expo.modules.notifications.notifications.interfaces.SchedulableNotificationTrigger
import expo.modules.notifications.notifications.model.NotificationContent
import expo.modules.notifications.notifications.model.NotificationRequest
import expo.modules.notifications.notifications.triggers.ChannelAwareTrigger
import expo.modules.notifications.notifications.triggers.DailyTrigger
import expo.modules.notifications.notifications.triggers.DateTrigger
import expo.modules.notifications.notifications.triggers.MonthlyTrigger
import expo.modules.notifications.notifications.triggers.TimeIntervalTrigger
import expo.modules.notifications.notifications.triggers.WeeklyTrigger
import expo.modules.notifications.notifications.triggers.YearlyTrigger
import expo.modules.notifications.service.NotificationsService
import expo.modules.notifications.service.NotificationsService.Companion.getAllScheduledNotifications
import expo.modules.notifications.service.NotificationsService.Companion.removeAllScheduledNotifications
import expo.modules.notifications.service.NotificationsService.Companion.removeScheduledNotification
import expo.modules.notifications.service.NotificationsService.Companion.schedule

open class NotificationScheduler : Module() {
  protected open val schedulingContext: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private val handler = Handler(Looper.getMainLooper())

  protected fun createResultReceiver(body: ResultReceiverBody) =
    createDefaultResultReceiver(handler, body)

  override fun definition() = ModuleDefinition {
    Name("ExpoNotificationScheduler")

    AsyncFunction("getAllScheduledNotificationsAsync") { promise: Promise ->
      getAllScheduledNotifications(
        schedulingContext,
        createResultReceiver { resultCode: Int, resultData: Bundle? ->
          if (resultCode == NotificationsService.SUCCESS_CODE) {
            val requests = resultData?.getParcelableArrayList<NotificationRequest>(NotificationsService.NOTIFICATION_REQUESTS_KEY)
            if (requests == null) {
              promise.reject("ERR_NOTIFICATIONS_FAILED_TO_FETCH", "Failed to fetch scheduled notifications.", null)
            } else {
              promise.resolve(serializeScheduledNotificationRequests(requests))
            }
          } else {
            val e = resultData?.getSerializable(NotificationsService.EXCEPTION_KEY) as Exception
            promise.reject("ERR_NOTIFICATIONS_FAILED_TO_FETCH", "Failed to fetch scheduled notifications.", e)
          }
        }
      )
    }

    AsyncFunction("scheduleNotificationAsync") { identifier: String, notificationContentMap: ReadableArguments, triggerParams: ReadableArguments?, promise: Promise ->
      try {
        val content = ArgumentsNotificationContentBuilder(schedulingContext).setPayload(notificationContentMap).build()
        val request = createNotificationRequest(
          identifier,
          content,
          triggerFromParams(triggerParams)
        )

        schedule(
          schedulingContext,
          request,
          createResultReceiver { resultCode: Int, resultData: Bundle? ->
            if (resultCode == NotificationsService.SUCCESS_CODE) {
              promise.resolve(identifier)
            } else {
              val e = resultData?.getSerializable(NotificationsService.EXCEPTION_KEY) as? Exception
              promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. ${e?.message}", e)
            }
          }
        )
      } catch (e: InvalidArgumentException) {
        promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. ${e.message}", e)
      } catch (e: NullPointerException) {
        promise.reject("ERR_NOTIFICATIONS_FAILED_TO_SCHEDULE", "Failed to schedule the notification. Encountered unexpected null value. ${e.message}", e)
      }
    }

    AsyncFunction("cancelScheduledNotificationAsync", this@NotificationScheduler::cancelScheduledNotificationAsync)

    AsyncFunction("cancelAllScheduledNotificationsAsync", this@NotificationScheduler::cancelAllScheduledNotificationsAsync)

    AsyncFunction("getNextTriggerDateAsync") { triggerParams: ReadableArguments?, promise: Promise ->
      try {
        val trigger = triggerFromParams(triggerParams)
        if (trigger is SchedulableNotificationTrigger) {
          val nextTriggerDate = trigger.nextTriggerDate()
          if (nextTriggerDate == null) {
            promise.resolve(null)
          } else {
            promise.resolve(nextTriggerDate.time.toDouble())
          }
        } else {
          val triggerDescription = if (trigger == null) "null" else trigger.javaClass.name
          val message = String.format("It is not possible to get next trigger date for triggers other than calendar-based. Provided trigger resulted in %s trigger.", triggerDescription)
          promise.reject("ERR_NOTIFICATIONS_INVALID_CALENDAR_TRIGGER", message, null)
        }
      } catch (e: InvalidArgumentException) {
        promise.reject("ERR_NOTIFICATIONS_FAILED_TO_GET_NEXT_TRIGGER_DATE", "Failed to get next trigger date for the trigger. ${e.message}", e)
      } catch (e: NullPointerException) {
        promise.reject("ERR_NOTIFICATIONS_FAILED_TO_GET_NEXT_TRIGGER_DATE", "Failed to get next trigger date for the trigger. Encountered unexpected null value. ${e.message}", e)
      }
    }
  }

  open fun cancelScheduledNotificationAsync(identifier: String, promise: Promise) {
    removeScheduledNotification(
      schedulingContext,
      identifier,
      createResultReceiver { resultCode: Int, resultData: Bundle? ->
        if (resultCode == NotificationsService.SUCCESS_CODE) {
          promise.resolve(null)
        } else {
          val e = resultData?.getSerializable(NotificationsService.EXCEPTION_KEY) as? Exception
          promise.reject("ERR_NOTIFICATIONS_FAILED_TO_CANCEL", "Failed to cancel notification.", e)
        }
      }
    )
  }

  open fun cancelAllScheduledNotificationsAsync(promise: Promise) {
    removeAllScheduledNotifications(
      schedulingContext,
      createResultReceiver { resultCode: Int, resultData: Bundle? ->
        if (resultCode == NotificationsService.SUCCESS_CODE) {
          promise.resolve(null)
        } else {
          val e = resultData?.getSerializable(NotificationsService.EXCEPTION_KEY) as? Exception
          promise.reject("ERR_NOTIFICATIONS_FAILED_TO_CANCEL", "Failed to cancel all notifications.", e)
        }
      }
    )
  }

  @Throws(InvalidArgumentException::class)
  protected fun triggerFromParams(params: ReadableArguments?): NotificationTrigger? {
    if (params == null) {
      return null
    }
    val channelId = params.getString("channelId", null)
    return when (val type = params.getString("type")) {
      "timeInterval" -> {
        val seconds = params["seconds"] as? Number
          ?: throw InvalidArgumentException("Invalid value provided as interval of trigger.")

        TimeIntervalTrigger(channelId, seconds.toLong(), params.getBoolean("repeats"))
      }

      "date" -> {
        val timestamp = params["timestamp"] as? Number
          ?: throw InvalidArgumentException("Invalid value provided as date of trigger.")

        DateTrigger(channelId, timestamp.toLong())
      }

      "daily" -> {
        val hour = params["hour"] as? Number
        val minute = params["minute"] as? Number

        if (hour == null || minute == null) {
          throw InvalidArgumentException("Invalid value(s) provided for daily trigger.")
        }

        DailyTrigger(
          channelId,
          hour.toInt(),
          minute.toInt()
        )
      }

      "weekly" -> {
        val weekday = params["weekday"] as? Number
        val hour = params["hour"] as? Number
        val minute = params["minute"] as? Number

        if (weekday == null || hour == null || minute == null) {
          throw InvalidArgumentException("Invalid value(s) provided for weekly trigger.")
        }
        WeeklyTrigger(
          channelId,
          weekday.toInt(),
          hour.toInt(),
          minute.toInt()
        )
      }

      "monthly" -> {
        val day = params["day"] as? Number
        val hour = params["hour"] as? Number
        val minute = params["minute"] as? Number

        if (day == null || hour == null || minute == null) {
          throw InvalidArgumentException("Invalid value(s) provided for yearly trigger.")
        }

        MonthlyTrigger(
          channelId,
          day.toInt(),
          hour.toInt(),
          minute.toInt()
        )
      }

      "yearly" -> {
        val day = params["day"] as? Number
        val month = params["month"] as? Number
        val hour = params["hour"] as? Number
        val minute = params["minute"] as? Number

        if (day == null || month == null || hour == null || minute == null) {
          throw InvalidArgumentException("Invalid value(s) provided for yearly trigger.")
        }

        YearlyTrigger(
          channelId,
          day.toInt(),
          month.toInt(),
          hour.toInt(),
          minute.toInt()
        )
      }

      "channel" -> ChannelAwareTrigger(channelId)
      else -> throw InvalidArgumentException("Trigger of type: $type is not supported on Android.")
    }
  }

  protected open fun createNotificationRequest(
    identifier: String,
    content: NotificationContent,
    notificationTrigger: NotificationTrigger?
  ): NotificationRequest {
    return NotificationRequest(identifier, content, notificationTrigger)
  }

  protected open fun serializeScheduledNotificationRequests(requests: Collection<NotificationRequest>): List<Bundle> {
    return requests.map(NotificationSerializer::toBundle)
  }
}
